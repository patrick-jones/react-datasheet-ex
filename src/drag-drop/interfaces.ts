import RDS from 'react-datasheet';
import {ConnectDragPreview, ConnectDragSource, ConnectDropTarget} from 'react-dnd';
import {ColumnHeader, HeaderRendererProps, RowRendererProps} from '../interfaces';

/**
 * Properties that must be provided to a drag and drop enabled row
 */
export interface DragDropRowProps<T extends RDS.Cell<T, V>, V = string> {
  /** The current row index */
  row: number;
  /** Callback for when a row is dropped on another */
  readonly onRowDropped: (source: number, target: number) => void;
}

/**
 * Properties that must be provided to a drag and drop enabled column header
 */
export interface DragDropHeaderProps {
  /** The current column index */
  col: number;
  /** Callback for one header is dropped on another */
  readonly onHeaderDropped: (source: number, target: number) => void;
}

/**
 * Drag source properties that are injected into a drag and drop provider
 */
export interface InjectedDragSourceProps {
  readonly connectDragSource: ConnectDragSource;
  readonly connectDragPreview: ConnectDragPreview;
}

/**
 * Drop target properties that are injected into a drag and drop provider
 */
export interface InjectedDropTargetProps {
  readonly connectDropTarget: ConnectDropTarget;
  readonly isDropOver: boolean;
}

/**
 * Combined drag source and drop target properties that are injected
 * into a drag and drop provider
 */
export interface InjectedDragDropProps extends InjectedDragSourceProps, InjectedDropTargetProps {}

/**
 * Extends `ReactDatasheet.RowRendererProps` to include
 * drag and drop properties
 */
export interface DragDropRowRendererProps<CellType, V = string>
  extends RowRendererProps<CellType, V>,
    DragDropRowProps<CellType, V> {}

/**
 * Extends `HeaderRendererProps` to include drag and drop properties.
 */
export interface DragDropHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends HeaderRendererProps<T>,
    DragDropHeaderProps {}
