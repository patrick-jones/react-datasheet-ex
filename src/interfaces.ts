import React, {CSSProperties} from 'react';
import RDS from 'react-datasheet';

export interface RowLocation {
  row: number;
}

export interface ColumnLocation {
  col: number;
}

export interface ColumnHeader {
  id: string;
  title: string;
  width?: number;
}

export interface HeaderRendererProps<T extends ColumnHeader = ColumnHeader> extends ColumnLocation {
  header: T;
}

export type HeaderRenderer<T extends ColumnHeader = ColumnHeader,
  R extends HeaderRendererProps<T> = HeaderRendererProps<T>> = React.ComponentType<R>;

export type HeaderContentRenderer<T extends ColumnHeader = ColumnHeader>
  = React.ComponentType<HeaderRendererProps<T>>;

export interface SheetRendererProps<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  >
  extends RDS.SheetRendererProps<T, V> {
  /** React style object that should be applied to the enclosing table */
  style?: CSSProperties | undefined;
  /**
   * An array of `ColumnHeader`, which are rendered into `th` elements
   * @default []
   */
  headers: H[];
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderRenderer
   */
  headerRenderer: HeaderRenderer;
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderContentRenderer
   */
  headerContentRenderer: HeaderContentRenderer;
}


export type ActionContentRenderer<T extends RowLocation = RowLocation>
  = React.ComponentType<T>;

export interface RowRendererProps<
    T extends RDS.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends RDS.RowRendererProps<T, V> {

  actionContentRenderer?: ActionContentRenderer<AP>;
}
