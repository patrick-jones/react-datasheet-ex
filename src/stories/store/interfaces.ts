import RDS from 'react-datasheet';
import {ColumnHeader, Overflow} from '../../interfaces';

export type ExampleCellValue = string | number | null;
export interface ExampleCellType extends RDS.Cell<ExampleCellType> {
  value: ExampleCellValue;
}

export interface ExampleModel {
  id: string;
  headers: ExampleHeader[];
  rows: ExampleRow[];
  selected: RDS.Selection | null;
  overflow: Overflow;
}

export interface ExampleHeader extends ColumnHeader {
  hidden?: boolean;
}

export interface ExampleRow {
  id: string;
  selected?: boolean;
  data: ExampleCellType[];
}

