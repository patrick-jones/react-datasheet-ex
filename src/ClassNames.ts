export const READ_ONLY_CELL = 'cell read-only';
export const ACTION_CELL = 'rdx-action-cell';

export const HEADER_CELL = 'rdx-header-cell';
export const HEADER_CELL_LEFT = 'rdx-header-cell--left';
export const HEADER_CELL_CENTER = 'rdx-header-cell--center';
export const HEADER_CELL_RIGHT = 'rdx-header-cell--right';

export const HEADER_CELL_CONTAINER = 'rdx-header-cell-container';
export const HEADER_CELL_TITLE = 'rdx-header-cell-title';

export const HEADER_DRAG_SOURCE = 'rdx-header-drag-source';
export const HEADER_DROP_TARGET = 'rdx-header-drop-target';

export const ROW = 'rdx-row';
export const ROW_DROP_TARGET = 'rdx-row-drop-target';
export const ROW_DRAG_SOURCE = 'rdx-row-drag-source';

export const RESIZING_WIDTH = 'rdx-ex-resizing-width';
export const DATA_GRID_RESIZABLE = 'rdx-data-grid-resizable';
export const WIDTH_RESIZE_HANDLE = 'rdx-width-resize-handle';

export const SCROLL_PANEL = 'rds-ex-scroll-panel';
