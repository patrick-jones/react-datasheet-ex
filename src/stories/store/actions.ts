import RDS from 'react-datasheet';
import {ExampleCellType} from './interfaces';
import {Overflow} from '../../interfaces';

/**
 * Some quick redux-like interfaces
 */

/**
 * A redux-like action
 */
export type Action<P = any> = {
  readonly type: string;
  readonly payload: P;
};

/**
 * Abstract type for a function that creates an action
 */
export type ActionCreator = (...args: any[]) => Action;

/**
 * Abstract type for a mapping of keys are ActionCreator functions
 */
export type ActionsCreators = {
  [key: string]: ActionCreator;
};

/**
 * Abstract type for the Action objects returned by an ActionCreators mapping
 */
export type Actions<A extends ActionsCreators> = ReturnType<A[keyof A]>;

/**
 * Type for actions once they are run through the dispatch function
 */
export type DispatchedActionCreator<AC extends ActionCreator> = (...args: any[]) => void;

/**
 * Converts `actions` object from functions that return an Action to
 * functions with the same arguments that return void.
 */
export type DispatchedActions<TAC> = {
  [K in keyof TAC]: TAC[K] extends ((...args: infer A) => Action) ? (...args: A) => void : undefined
};

const createAction = <T extends string, P>(type: T, payload: P) => ({type, payload});

/**
 * Our specific actions
 */
export enum ActionTypes {
  CELLS_CHANGED = 'rds-ex__cells-changed',
  SELECTION_CHANGED = 'rds-ex__selection-changed',
  OVERFLOW_CHANGED = 'rds-ex__overflow-changed',
  HEADER_DROPPED = 'rds-ex__header-dropped',
  ROW_DROPPED = 'rds-ex__row-dropped',
  ROW_SELECTION_CHANGED = 'rds-ex__row-selection-changed',
  COLUMN_HIDDEN_CHANGED = 'rds-ex__column-hidden-changed',
  HEADER_RESIZING = 'rds-ex__header-resizing',
  HEADER_RESIZE_END = 'rds-ex__header-resize-end',
}

export const SheetActionCreators = {
  cellsChanged:
    (changes: RDS.CellsChangedArgs<ExampleCellType>, additions?: RDS.CellAdditionsArgs<string>) =>
      createAction(ActionTypes.CELLS_CHANGED, {changes, additions}),
  selectionChanged: (selected: RDS.Selection) =>
    createAction(ActionTypes.SELECTION_CHANGED, {selected}),
  overflowChanged: (overflow: Overflow) =>
    createAction(ActionTypes.OVERFLOW_CHANGED, {overflow}),
  headerDropped: (source: number, target: number) =>
    createAction(ActionTypes.HEADER_DROPPED, {source, target}),
  rowDropped: (source: number, target: number) =>
    createAction(ActionTypes.ROW_DROPPED, {source, target}),
  rowSelectionChanged: (row: number, selected: boolean) =>
    createAction(ActionTypes.ROW_SELECTION_CHANGED, {row, selected}),
  columnHiddenChanged: (id: string, hidden: boolean) =>
    createAction(ActionTypes.COLUMN_HIDDEN_CHANGED, {id, hidden}),
  headerResizing: (source: number, requestedSize: number) =>
    createAction(ActionTypes.HEADER_RESIZING, {source, requestedSize}),
  headerResizeEnd: (source: number, requestedSize: number) =>
    createAction(ActionTypes.HEADER_RESIZE_END, {source, requestedSize}),
};

export type SheetActionCreators = typeof SheetActionCreators;
