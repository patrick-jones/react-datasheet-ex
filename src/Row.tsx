import React, {Component, ReactNode} from 'react';
import RDS from 'react-datasheet';

import {READ_ONLY_CELL, ACTION_CELL} from './ClassNames';
import DefaultAction from './DefaultAction';
import {ActionRendererProps, RowRendererProps} from './interfaces';


const rowClass = `${READ_ONLY_CELL} ${ACTION_CELL}`;


export default class Row<
  T extends RDS.Cell<T, V>, V = string,
  ARP extends ActionRendererProps<T, V> = ActionRendererProps<T, V>,
> extends Component<RowRendererProps<T, V, ARP>> {

  static displayName = 'RowRenderer';

  static readonly defaultProps = {
    actionRenderer: DefaultAction,
  };

  render(): ReactNode {
    const {id, row, children, actionRenderer: ActionRenderer, ...rest} = this.props;
    return (
      <tr key={`$row-${id || row}`}>
        <td key='$$actionCell' className={rowClass}>
          <ActionRenderer id={id} row={row} {...rest} />
        </td>
        {children}
      </tr>
    );
  }
}
