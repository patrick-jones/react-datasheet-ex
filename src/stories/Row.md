# Row

This is a companion to `Sheet`. Like `Sheet` it is not strictly required but
it will make things easier. Each row renders with an extra read-only cell 
at the beginning. This "action" cell does not do anything by itself, but is 
used as the drag handle for draggable rows and can optionally house other controls
such as a check box or row labels. This example uses the action cell to support 
row selection.


<!-- STORY -->

## Prop Types

### Row

This component differs from the default renderer in `react-datasheet` in that it
adds a blank cell at the beginning of each row. This extra cell is used as the
drag handle 
 
In addition to the
[standard `rowRenderer` properties](https://github.com/nadbm/react-datasheet#row-renderer)
supplied by `react-datasheet`, this component can accept a custom renderer for
the action cell's content.

Name | Type | Description
:--- | :--- | :---
row             | number | The 0-based row index
cells           | Array  | The cells in the current row
id?             | string | Optional unique (within the sheet) id for the row. If provided, it will be used as the row's `key` attribute, since using indices for keys is [not recommended](https://reactjs.org/docs/lists-and-keys.html#keys)
actionRenderer? | `ActionRenderer` | Optional. A function or component that renders the content of the action cell; default is to leave it empty 


### type ActionRenderer

A component or function that renders the row action cell content. It takes the
same properties as `Row` minus `actionRenderer`. 

Name | Type | Description
:--- | :--- | :---
row             | number | The 0-based row index
cells           | Array  | The cells in the current row
id?             | string | Optional unique (within the sheet) id for the row.


## Sample code
```js
import React, {Component} from 'react';
import ReactDatasheet from 'react-datasheet';
import memoize from 'memoize-one';

import {Row, Sheet} from 'react-datasheet-ex';
import {renderValue, rowsToData} from './store/model';

export default class RowExample extends Component {

  getData = memoize(rowsToData);

  sheetRenderer = props => {
    const {model: {headers, overflow}} = this.props;

    return (
      <Sheet
        headers={headers}
        overflow={overflow}
        {...props}
      />
    );
  }

  rowRenderer = props => {
    return (
      <Row
        actionRenderer={this.actionRenderer}
        {...props}
      />
    );
  }

  actionRenderer = props => {
    const {actions, model} = this.props;
    const {row} = props;
    const {selected} = model.rows[props.row];
    return (
      <input
        type='checkbox'
        checked={selected}
        onClick={() => actions.rowSelectionChanged(row, !selected)}
      />
    );
  }

  render() {

    const {actions, model: {rows, selected}} = this.props;

    return (
      <ReactDatasheet
        data={this.getData(rows)}
        sheetRenderer={this.sheetRenderer}
        rowRenderer={this.rowRenderer}
        valueRenderer={renderValue}
        onCellsChanged={actions.cellsChanged}
        onSelect={actions.selectionChanged}
        selected={selected}
      />
    );
  }
}

```
