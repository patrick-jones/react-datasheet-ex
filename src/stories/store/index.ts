import {model} from './model';
import {ExampleModel} from './interfaces';
import {Actions, SheetActionCreators} from './actions';
import reducer from './reducer';

export {model};

export function preprocess(model: ExampleModel, ...actions: Actions<SheetActionCreators>[]) {
  return actions.reduce(
    (reply, action) => {
      const nextState = reducer(reply, action);
      return Object.assign({}, reply, nextState);
    },
    model
  );
}
