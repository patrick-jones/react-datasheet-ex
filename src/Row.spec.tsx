import React from 'react';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';

import {ACTION_CELL} from './ClassNames';
import Row from './Row';
import {ExampleCellType} from './stories/store/interfaces';

const cells: ExampleCellType[] = [
  {value: 'cell'},
];

describe('Row basic rendering', () => {
  it('Should render', () => {
    const wrapper = mount(
      <table>
        <tbody>
          <Row row={0} cells={cells}>
            <td className='cell'>cell</td>
          </Row>
        </tbody>
      </table>
    );

    const td = wrapper.find('td.cell');
    expect(td.length).toBe(2);
    expect(td.find(`td.${ACTION_CELL}`).length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should render custom action content', () => {
    const wrapper = mount(
      <table>
        <tbody>
        <Row
          row={0}
          cells={cells}
          actionRenderer={() => (<span className='abcd-puppy'>Action</span>)}
        >
          <td className='cell'>cell</td>
        </Row>
        </tbody>
      </table>
    );

    expect(wrapper.find(`td.${ACTION_CELL} span.abcd-puppy`).text())
      .toBe('Action');
  });
});
