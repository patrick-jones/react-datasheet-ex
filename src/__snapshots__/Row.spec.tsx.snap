// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row basic rendering Should render 1`] = `
<table>
  <tbody>
    <RowRenderer
      actionRenderer={[Function]}
      cells={
        Array [
          Object {
            "value": "cell",
          },
        ]
      }
      row={0}
    >
      <tr>
        <td
          className="cell read-only rdx-action-cell"
          key="$$actionCell"
        >
          <DefaultAction
            cells={
              Array [
                Object {
                  "value": "cell",
                },
              ]
            }
            row={0}
          />
        </td>
        <td
          className="cell"
        >
          cell
        </td>
      </tr>
    </RowRenderer>
  </tbody>
</table>
`;
